http: improve connection lifecycle management

nvartolomei opened the pull request:
[Copilot summary](https://github.com/redpanda-data/redpanda/pull/28788#pullrequestreview-3519748268)

> This PR improves HTTP client connection handling and reuse by properly managing keep-alive connections and detecting when connections should be closed. The changes ensure the client properly invalidates itself when responses are not fully consumed or when the server explicitly requests connection closure via the "Connection: close" header.
> 
> Key changes:
> 
> - Added destructor logic to response_stream to detect improper usage and server-requested connection closure
> - Added support for HTTP HEAD requests by configuring the parser to skip the response body
> - Exposed the is_valid() method from base_transport to allow checking connection validity
> 

  See individual commits for details.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Improvements

* Minor improvements to http client for better connection reuse in certain edge cases.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
http: drop unused utils dep

nvartolomei committed:
http: include http verb in the logger

nvartolomei committed:
http: respect connection close directive

Notably AWS S3 always sends `Connection: close` in response to
_DeleteObjects_ operation. Since we didn't respect it the request we
attempt on the same connection always fails resulting in backoff kicking
in (if retries are allowed) resulting in unnecessary delay.

Fortunately, since our http client always creates a `response_stream`
referencing a client we can hook into the destructor and conditionally
invalidate the client/close our connection end. This way, when we
attempt to reuse it we know that we need to create a connection first.

You can see the problem yourself using rpk self-test with AWS S3.

Separately we should consider setting backoff to 0 for the first retry.

nvartolomei committed:
http: consume body for HEAD responses

Although the HTTP RFCs require servers not to send any body in the
responses we can't assume that all servers are compliant. So we are
defensive and always _parse_ the body. However, we instruct boost::beast
to skip over it. This guarantees a reusable client/connection.

The actual driver is not non-compliant servers but rather leaving
boost::beast parser in a valid state, always. Before this change, after
handling HEAD responses, the parses would have been left in an
incomplete state.

This change unlocks an improvement of the resilience of our http client
in face of bugs which we'll implement next. I.e. invalidate clients left
in an _improper_ state.

nvartolomei committed:
http: invalidate client on incomplete response consumption

Ensures dirty clients are not reused by shutting them down when response
streams are destroyed before the response is fully consumed.

dotnwat commented on src/v/http/client.cc:
```diff
@@ -316,6 +316,24 @@ client::response_stream::response_stream(
+          _ctxlog.trace,
+          "response_stream indicates connection close via header, "
+          "shutting down the client");
+        _client->shutdown();
```
shutting down the client in the destructor here makes me slightly uneasy from a concurrency persepective w.r.t. things like transport::wait_input_shutdown and transport::stop since afaik our shutdown interfaces are generally called explicitly and so we can necessarily observe their order.

maybe my concern is unjustified, but if not, maybe we'd want to the response stream destructor to set a flag or something on the client and then let other interactions with the client inject the shutdown?

Lazin replied:
This is similar to a flag. The `shutdown` method just resets few fields so the subsequent call to `get_connected` on the same http client would connect. It should be safe. The `shutdown` is invoked from many places. There is a guarantee that the same client will not be used by several `response_stream` objects.

dotnwat replied:
thanks that makes sense

dotnwat commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -835,8 +835,8 @@ ss::future<s3_client::head_object_result> s3_client::do_head_object(
-            return ref->prefetch_headers().then(
-              [ref, key]() -> ss::future<head_object_result> {
+            return util::drain_response_stream(ref).then(
+              [ref, key](const iobuf&) -> ss::future<head_object_result> {
```
a little confused about this change. IIUC for a head request we'd want to still read and parse the headers, then your observation that we should then skip the body to account for non-compliant servers makes sense, but drain_response_stream doesn't seem to process the headers?

> So we are
defensive and always _parse_ the body. However, we instruct boost::beast
to skip over it.

just to be clear, i don't think boost::beast (or anything else) would be parsing the body, right? the only parsing should be the http headers?

Lazin replied:
This will parse status code and headers. Both `prefetch_headers` and `drain_response_stream` have similar implementation. The difference is that `prefetch_headers` is consuming from the stream until all headers are ready. The `drain_*` method on the other hand consumes until the EOF unconditionally. The main problem with it is that it checks the headers to find if the chunked transfer encoding is enabled and then it uses different implementations for normal and chunked encodings. But here we will be checking headers before the header is done. I think that the better alternative is to use `http::drain` here. And the current implementation is correct if we assume that the cloud storage provider will never reply with chunked transfer encoded to the HeadObject request which is a valid assumption to make.

nvartolomei replied:
> But here we will be checking headers before the header is done. 

I'm not following. What do you mean we are checking headers before it is done?

dotnwat replied:
the only interpretation of @Lazin's comment i can come up with is that drain_* reads and parses the headers up to EOF (what we want), but it requires knowing chunked/non-chunked before it starts (that is, it doesn't adapt to the headers it just throws them away) as some kind of out-of-band information provided by the caller.

i think we probably want something more generic here (e.g. http::drain as suggested)

nvartolomei replied:
To me both actually look as equivalent 🤔 i.e. `drain_chunked_response_stream` doesn't do anything special 🤔 and should be removed. I'm reading it wrong?

dotnwat replied:
>  I'm reading it wrong?

not sure. i'm just going off what I think @lazin is telling us

Lazin replied:
> I'm not following. What do you mean we are checking headers before it is done?

In this method there is a check that depends on headers which are not arrived yet, it's expected that you will receive headers first and then invoke this method.

Lazin replied:
> To me both actually look as equivalent 🤔 i.e. `drain_chunked_response_stream` doesn't do anything special 🤔 and should be removed. I'm reading it wrong?

it's doing the same thing, my point was that it doesn't look correct to check headers before they're arrived, so maybe this function need some refactor

nvartolomei replied:
- Removed duplicate implementations of draining.
- Added invariant check with an exception.

This should be much better. I'll re-request reviews once I get a green CI. 🤞

dotnwat reviewed (commented):
lgtm, just some clarifying questions. i think we would benefit from @Lazin taking a peek too.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--2b1819fe-36e6-42d3-b940-2480fdea0a2d-->
<!--TEST RESULTS-->
<!--f05bbc26-cf77-413d-b592-148081c2d01f-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77112">build#77112</a></summary>


| test_class              | test_method             | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                        |
|:------------------------|:------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------|
| ReplicatedMetastoreTest | TestBasicRemoveTopics   |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/77112#019acbe9-3862-4034-a33f-2e37644e372e | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                     |
| ScalingUpTest           | test_fast_node_addition | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77112#019acc25-1818-415e-bb8d-b7a7676b6a7c | FLAKY         | 20/21    | upstream reliability is '98.51851851851852'. current run reliability is '95.23809523809523'. drift is 3.28042 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77184">build#77184</a></summary>


| test_class                   | test_method           | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                  |
|:-----------------------------|:----------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ReplicatedMetastoreTest      | TestBasicRemoveTopics |                           | unit        | https://buildkite.com/redpanda/redpanda/builds/77184#019adfa2-82bb-4648-a61c-d291272b8b88 | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                               |
| MountUnmountIcebergTest      | test_simple_remount   | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/77184#019adfe1-1a3f-4e0f-b32b-536e3c12db2b | FLAKY         | 15/21    | upstream reliability is '79.52488687782805'. current run reliability is '71.42857142857143'. drift is 8.09632 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount     |
| PartitionMovementUpgradeTest | test_basic_upgrade    | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/77184#019adfe1-1a42-4507-b5ea-634606ad33c8 | FLAKY         | 20/21    | upstream reliability is '99.44289693593315'. current run reliability is '95.23809523809523'. drift is 4.2048 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementUpgradeTest&test_method=test_basic_upgrade |
</details>
<!--/body-->

nvartolomei commented:
/cdt
tests/rptest/fuzz/random_node_operations_test.py
tests/rptest/scale_tests/tiered_storage_io_stress_test.py

dotnwat reviewed (commented):
@lazin is it better to use `http::drain` than to use an interface that makes assumptions about encoding just because it's s3? for example, it would seem that `http::drain` could possibly be used in a generic way rather than an s3 specific wya?

nvartolomei commented:
Oh. We definitely should fix that and add an assert/exception.

On Mon, 1 Dec 2025 at 22:12, Eugene Lazin ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/v/cloud_storage_clients/s3_client.cc
> <https://github.com/redpanda-data/redpanda/pull/28788#discussion_r2578904735>
> :
>
> > -            return ref->prefetch_headers().then(
> -              [ref, key]() -> ss::future<head_object_result> {
> +            return util::drain_response_stream(ref).then(
> +              [ref, key](const iobuf&) -> ss::future<head_object_result> {
>
> To me both actually look as equivalent 🤔 i.e.
> drain_chunked_response_stream doesn't do anything special 🤔 and should
> be removed. I'm reading it wrong?
>
> it's doing the same thing, my point was that it doesn't look correct to
> check headers before they're arrived, so maybe this function need some
> refactor
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redpanda-data/redpanda/pull/28788#discussion_r2578904735>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEETWMK3FXXXUUWHGGOLY337S4LNAVCNFSM6AAAAACNP524YGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTKMRXGI3TGOJZGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

nvartolomei committed:
http: add invariant check to get_headers

Prevent interface misuse.

nvartolomei committed:
http/tests: use http::drain instead of hand rolled version

nvartolomei committed:
http/tests: more robust chunked encoding test

nvartolomei committed:
cloud_storage_clients: replace custom draining with http::drain

Their functionality is equivalent if you look close enough. Old code is
extra confusing for absolutely no good reason.

nvartolomei committed:
http: implement draining without accumulator

dotnwat commented on src/v/http/tests/http_client_test.cc:
```diff
@@ -648,7 +648,10 @@ SEASTAR_THREAD_TEST_CASE(test_http_via_impostor_chunked_encoding) {
     // Generate response
-    http::chunked_encoder encoder{false};
+    constexpr size_t http_body_chunk_size = 16;
+    // Ensure we have data for more than one chunk for a more robust test.
```
🥇

dotnwat commented on src/v/http/client.cc:
```diff
@@ -327,6 +327,11 @@ bool client::response_stream::is_header_done() const {
 /// Access response headers (should only be called if is_header_done() == true)
 const client::response_header& client::response_stream::get_headers() const {
+    if (!is_header_done()) {
+        throw std::runtime_error(
```
if it's an invariant, what's holding us back from using vassert?

nvartolomei replied:
My rule for vasserts is that they are fine when I have high confidence someone won't hit it by mistake. I.e. internal invariants.
For invariants that are hard to prove/have confidence about (i.e. public APIs) I prefer exceptions. Especially if I can't prove that code paths don't have coverage during normal runs but may be hit in some edge cases. Example of that is some of Azure adlsv2 code paths. Don't want to crash redpanda for something that not necessarily a core part of the system.

dotnwat replied:
i think we should add this to our coding guidelines

dotnwat commented on src/v/cloud_storage_clients/util.cc:
```diff
@@ -150,52 +150,6 @@ handle_client_transport_error<ss::logger>(std::exception_ptr, ss::logger&);
-ss::future<iobuf>
-drain_chunked_response_stream(http::client::response_stream_ref resp) {
-    return ss::do_with(
-      resp->as_input_stream(),
```
hmm, `response_data_source` does vaguely look equivalent to the is_done/recv_some loop structure. sure is some fine tuning of the buffer in there trimming and what not :/

dotnwat commented on src/v/http/client.h:
```diff
@@ -360,8 +360,16 @@ redacted_fields();
 ss::future<boost::beast::http::status>
 status(client::response_stream_ref response);
 
+template<typename T = iobuf>
+requires std::same_as<T, iobuf> || std::same_as<T, void>
+ss::future<T> drain(client::response_stream_ref response);
+
+template<>
 ss::future<iobuf> drain(client::response_stream_ref response);
 
+template<>
+ss::future<> drain(client::response_stream_ref response);
```
maybe `drain` and `drain_to_iobuf`? or like, is the template stuff providing any value here?

nvartolomei replied:
Looks neat with a template :)

dotnwat replied:
i'm sure there is an xkcd for this

Lazin replied:
I agree with using a different name for this

nvartolomei replied:
Let's experiment with leaving it like this and see what experience will tell.

dotnwat replied:
> Let's experiment with leaving it like this and see what experience will tell.

the outcome of the experiment is already known: we'll use the templated version because the right time to change it was before merging and now its not worth changing.

dotnwat commented on src/v/http/client.cc:
```diff
@@ -314,6 +314,7 @@ client::response_stream::response_stream(
   , _sprobe(client->_probe->create_request_subprobe(convert_to_pverb(v))) {
     _parser.body_limit(std::numeric_limits<uint64_t>::max());
     _parser.eager(true);
+    _parser.skip(v == client::response_stream::verb::head);
```
does this have any affect for us?

nvartolomei replied:
It does. Without this change the beast parser would always expect a body part but we wouldn't have it for most HEAD responses. This is pre-requisite/requirement for the next commit in the chain https://github.com/redpanda-data/redpanda/pull/28788/commits/c03592390ed0551ce55e9bb2579464e9637627b6

dotnwat replied:
> body part but we wouldn't have it for most HEAD responses

this i don't understand. the body part is just the data after the header, right? what does boost care, it's not parsing that anyway right?

nvartolomei replied:
The entire response is piped via beast parser and the internal state machine is driven by it.

dotnwat replied:
weird. what does beast care about the body? maybe if it is chunked encoding?

nvartolomei replied:
Chunked encoding. Also finding the end of the body i.e. `\r\n` and telling us when to stop.

dotnwat replied:
i assume that `\r\n` as an ending to the body is specific (e.g. `0\r\n` as the final part of the body). for non-chunked encoding there is no explicit ending marker?

nvartolomei replied:
🤦 Right... For non-chunked it is done after reading Content-Length bytes of the body.

nvartolomei replied:
I'm re-reading this and now I question why it is needed too https://github.com/boostorg/beast/blob/aefc564fa72c07a3afc8a4cc8b460476f3c35a35/include/boost/beast/http/impl/basic_parser.ipp#L418

dotnwat replied:
all good. it probably doesn't matter for us, but didn't understand what it does so was also curious if we didn't want it.

nvartolomei replied:
@dotnwat apologies for previous link, that was for request parsing. Relevant code for response parsing is here: https://github.com/boostorg/beast/blob/aefc564fa72c07a3afc8a4cc8b460476f3c35a35/include/boost/beast/http/impl/basic_parser.ipp#L465-L520

Relevant bits:

```cpp
finish_header(...)
{
    // RFC 7230 section 3.3
    // https://tools.ietf.org/html/rfc7230#section-3.3

    // ===> nv: This is the branch we want to take <===
    if( (f_ & flagSkipBody) ||  // e.g. response to a HEAD request
        status_  / 100 == 1 ||   // 1xx e.g. Continue
        status_ == 204 ||        // No Content
        status_ == 304)          // Not Modified
    {
        // VFALCO Content-Length may be present, but we
        //        treat the message as not having a body.
        //        https://github.com/boostorg/beast/issues/692
        state_ = state::complete;
    }

    // ===> nv: But this is the branch we take without flagSkipBody <===
    // ===>     That is because Content-Length for head responses   <===
    // ===>     is greater than zero if GET request for the same    <===
    // ===>     URI would have returned content. However, the       <===
    // ===>     content is missing for the HEAD response we are     <===
    // ===>     processing.                                         <===
    else if(f_ & flagContentLength)
    {
        if(len_ > 0)
        {
            f_ |= flagHasBody;
            state_ = state::body0;
        ...
    }
    else ...
```

nvartolomei replied:
From RFC 7230:

> Responses to the HEAD request method (Section 4.3.2
   of [RFC7231]) never include a message body because the associated
   response header fields (e.g., Transfer-Encoding, Content-Length,
   etc.), if present, indicate only what their values would have been if
   the request method had been GET ([Section 4.3.1 of [RFC7231]](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.1)).

dotnwat replied:
nice, thanks!

dotnwat reviewed (approved):
lgtm. would like to have @Lazin take a final look

Lazin commented on src/v/http/client.cc:
```diff
@@ -783,4 +784,11 @@ ss::future<iobuf> drain(client::response_stream_ref response) {
 }
 
+template<>
+ss::future<> drain(client::response_stream_ref response) {
```
how is this correct? these are two overloads with the same input param but different return results? or I understand this incorrectly?

nvartolomei replied:
My understanding is that they are actually 2 different functions in this case rather than overloaded function.

dotnwat replied:
yeh i think the templating makes them different (i also had the same question)

Lazin reviewed (approved):
This looks good. The `http::drain` thing is a nit.

vbotbuildovich commented:
/backport v25.3.x
